influxdb3_write/src/parquet_cache/mod.rs

```diff
@@ -108,59 +125,213 @@ struct CacheValue {
+            }
+        }
     }
 
-    fn keep(&self) -> bool {
-        self.is_fetching() || self.is_success()
+    /// Check if an entry in the cache is in process of being fetched or if it was already fetched
+    /// successfully
+    fn path_already_fetched(&self, path: &Path) -> bool {
+        self.map.get(path).is_some()
     }
-}
 
-/// Implements the [`WeightScale`] trait to determine a [`CacheEntry`]'s size on insertion to
-/// the cache
-#[derive(Debug)]
-struct CacheEntryScale;
+    /// Insert a `Fetching` entry to the cache with a watcher
+    fn set_fetching(&self, path: &Path) {
+        let (tx, _) = watch::channel(None);
+        let entry = CacheEntry {
+            state: CacheEntryState::Fetching(tx),
+            hit_time: AtomicI64::new(self.time_provider.now().timestamp_nanos()),
+        };
+        let additional = entry.size();
+        self.map.insert(path.clone(), entry);
+        self.used.fetch_add(additional, Ordering::SeqCst);
+    }
+
+    /// Update a `Fetching` entry to a `Success` entry in the cache
+    fn set_success(&self, path: &Path, value: CacheValue) -> Result<(), anyhow::Error> {
+        match self.map.entry(path.clone()) {
+            Entry::Occupied(mut o) => {
+                let entry = o.get_mut();
+                let tx = match &entry.state {
+                    CacheEntryState::Fetching(tx) => tx.clone(),
+                    _ => bail!("attempted to set success state on a non-fetching cache entry"),
+                };
+                let cache_value = Arc::new(value);
+                entry.state = CacheEntryState::Success(Arc::clone(&cache_value));
+                entry
+                    .hit_time
+                    .store(self.time_provider.now().timestamp_nanos(), Ordering::SeqCst);
+                // TODO(trevor): what if size is greater than cache capacity?
+                let additional = entry.size();
+                self.used.fetch_add(additional, Ordering::SeqCst);
+                // notify any watching requests that tried to get this entry while it was
+                // fetching:
+                let _ = tx.send(Some(cache_value));
+                Ok(())
+            }
+            Entry::Vacant(_) => bail!("attempted to set success state on an empty cache entry"),
+        }
+    }
+
+    /// Remove an entry from the cache, as well as its associated size from the used capacity
+    fn remove(&self, path: &Path) {
+        let Some((_, entry)) = self.map.remove(path) else {
+            return;
+        };
+        self.used.fetch_sub(entry.state.size(), Ordering::SeqCst);
+    }
+
+    /// A cache should be pruned if it is using all of its memory capacity
+    fn should_prune(&self) -> bool {
+        let used = self.used.load(Ordering::SeqCst);
+        used >= self.capacity
+    }
+
+    /// Prune least recently hit entries from the cache
+    fn prune(&self) {
+        let n_to_prune = (self.map.len() as f64 * self.prune_percent).floor() as usize;
```

Comment:
It decides to prune using the `should_prune` method which only returns `true` when the cache has exceeded its capacity, but I suppose that could be called internally by this function instead, to make that decision more clear.

Reply:
I think I see what you are saying in terms of pruning entries that are only nanoseconds apart. I suppose we could have a different rule that prunes after a certain age, i.e., maybe there would actually be more than 10% that are suitable to be pruned, because they have aged out, and we might as well prune them while we're at it... but I don't know that I am up for tackling that on this PR.

Reply:
Yes sure - it was more to check if it was done intentionally that way. I think once we see the access patterns I guess we'd know more on how to fine tune the results.